src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -433,14 +431,19 @@ cdef class Polynomial(CommutativePolynomial):
             ...
             TypeError: keys do not match self's parent
         """
-        if len(x) == 1 and isinstance(x[0], dict):
-            g = self._parent.gen()
-            if g in x[0]:
-                return self(x[0][g])
-            elif len(x[0]) > 0:
-                raise TypeError("keys do not match self's parent")
-            return self
-        return self(*x, **kwds)
+        if not in_dict:
+            return self(*args, **kwds)
```

Comment:
Note that this is now a subtle change in behavior. Before, the empty dict `{}` would perform no substitutions if extra arguments were passed. Now it will do a substitution based on the inputs. The behavior is not well-defined in either case (mixing two input methods), but this will now agree with the multivariate case. I am just pointing this out this behavior change.

Reply:
I agree.  Note that, a priori, `subs` is not disallowing mixing of substitution methods:
```
sage: R.<x,y> = QQ[]
sage: p = x + y
sage: p.subs({0: 1}, y=2)
3
```